tools/graphiteindex/src/main/java/com/criteo/biggraphite/graphiteindex/MetricsIndex.java

```diff
@@ -0,0 +1,262 @@
+    }
+
+    public List<Pair<String, Long>> searchPaths(String pattern)
+    {
+        return search(
+            pattern,
+            doc -> Pair.of(
+                MetricPath.getPathFromDocument(doc),
+                MetricPath.getOffsetFromDocument(doc)
+            )
+        );
+    }
+
+    private <T> List<T> search(String pattern, Function<Document, T> handler)
+    {
+        BooleanQuery query = patternToQuery(pattern);
+        logger.trace("{} - Searching for '{}', generated query: {}", name, pattern, query);
+
+        ArrayList<T> results = new ArrayList<>();
+        Collector collector = new MetricsIndexCollector(
+            doc -> results.add(handler.apply(doc))
+        );
+
+        try {
+            IndexSearcher searcher = searcherMgr.acquire();
+            try {
+                // TODO(d.forest): we should probably use TimeLimitingCollector
+                searcher.search(query, collector);
+            } catch(IOException e) {
+                logger.error("{} - Cannot finish search query", name, e);
+            } finally {
+                searcherMgr.release(searcher);
+            }
+        } catch(IOException e) {
+            logger.error("{} - Cannot acquire index searcher", name, e);
+        }
+
+        return results;
+    }
+
+    /**
+     * Generates a Lucene query which matches the same metrics the given Graphite
+     * globbing pattern would.
+     */
+    private BooleanQuery patternToQuery(String pattern)
+    {
+        BooleanQuery.Builder queryBuilder = new BooleanQuery.Builder();
+
+        int length = MetricPath.iterateOnElements(
+            pattern,
+            (element, depth) -> {
+                // Wildcard-only fields do not filter anything
+                // TODO(d.forest): detect several wildcards?
+                if (element == "*") {
+                    return;
+                }
+
+                String termName = MetricPath.FIELD_PART_PREFIX + depth;
+                Query query;
+                if (element.indexOf('{') != -1 || element.indexOf('[') != -1) {
+                    query = new RegexpQuery(
+                        new Term(termName, graphiteToRegex(element)),
+                        RegExp.NONE
+                    );
+                } else if (element.indexOf('*') != -1 || element.indexOf('?') != -1) {
+                    query = new WildcardQuery(new Term(termName, element));
+                } else {
+                    query = new TermQuery(new Term(termName, element));
+                }
+
+                queryBuilder.add(query, BooleanClause.Occur.MUST);
+            }
+        );
+
+        queryBuilder.add(
+            IntPoint.newExactQuery(MetricPath.FIELD_LENGTH, length),
+            BooleanClause.Occur.MUST
+        );
+
+        return queryBuilder.build();
+    }
+
+    /**
+     * Naively transforms a Graphite globbing pattern into a regular expression.
+     */
+    private String graphiteToRegex(String query)
+    {
```

Comment:
Maybe add a todo with a pointer here :)

Reply:
AFAIK it supports the whole shebang, albeit without parsing/checking